### ISSUE TITLE
Datatype changes after recent addition on model files

### ISSUE THREAD

**Issue by jonespm** [tl-its-umich-edu/my-learning-analytics#495]
I ran [liquibase](https://www.liquibase.org/) on the database after the addition of the Django migrations and many datatypes were changed. The file_access was noted in #494 but we should either re-set the other columns to match what used to be there or write and run a manual migration.

```
Compared Schemas: my_learning_analytics_dev -> student_dashboard
Product Name: EQUAL
Product Version:
     Reference:   '5.7.17-log'
     Target: '5.7.22'
Missing Column(s):
     my_learning_analytics_dev.file_access.file_id
     my_learning_analytics_dev.file_access.user_id
Unexpected Column(s):
     student_dashboard.file_access.file_id_id
     student_dashboard.file_access.id
     student_dashboard.file_access.user_id_id
Changed Column(s):
     my_learning_analytics_dev.file_access.access_time
          defaultValue changed from 'CURRENT_TIMESTAMP' to 'null'
          order changed from '3' to '2'
          type changed from 'TIMESTAMP' to 'DATETIME'
     my_learning_analytics_dev.assignment.assignment_group_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.submission.assignment_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.submission.avg_score
          type changed from 'FLOAT(12)' to 'DOUBLE(22)'
     my_learning_analytics_dev.academic_terms.canvas_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.course.canvas_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment_weight_consideration.consider_weight
          defaultValue changed from 'false' to 'null'
     my_learning_analytics_dev.assignment.course_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment_groups.course_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment_weight_consideration.course_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.course_view_option.course_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.file.course_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.submission.course_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.user.course_id
          nullable changed from 'false' to 'true'
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.user_default_selection.course_id
          nullable changed from 'false' to 'true'
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.user.current_grade
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.academic_terms.date_end
          defaultValue changed from 'CURRENT_TIMESTAMP' to 'null'
          nullable changed from 'false' to 'true'
          type changed from 'TIMESTAMP' to 'DATETIME'
     my_learning_analytics_dev.academic_terms.date_start
          defaultValue changed from 'CURRENT_TIMESTAMP' to 'null'
          nullable changed from 'false' to 'true'
          type changed from 'TIMESTAMP' to 'DATETIME'
     my_learning_analytics_dev.user_default_selection.default_view_type
          nullable changed from 'false' to 'true'
          type changed from 'CHAR(225 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.user_default_selection.default_view_value
          type changed from 'CHAR(225 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment_groups.drop_highest
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment_groups.drop_lowest
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment.due_date
          type changed from 'TIMESTAMP' to 'DATETIME'
     my_learning_analytics_dev.user.final_grade
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.submission.graded_date
          type changed from 'TIMESTAMP' to 'DATETIME'
     my_learning_analytics_dev.assignment_groups.group_points
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.academic_terms.id
          type changed from 'BIGINT(20) UNSIGNED' to 'BIGINT(19)'
     my_learning_analytics_dev.assignment.id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment_groups.id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.course.id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.file.id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.submission.id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.user.id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment.local_date
          type changed from 'TIMESTAMP' to 'DATETIME'
     my_learning_analytics_dev.academic_terms.name
          defaultValue changed from '' to 'null'
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment.name
          defaultValue changed from '' to 'null'
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment_groups.name
          defaultValue changed from '' to 'null'
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.course.name
          defaultValue changed from '' to 'null'
          order changed from '4' to '3'
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.file.name
          type changed from 'TEXT(65535)' to 'LONGTEXT(2147483647)'
     my_learning_analytics_dev.user.name
          defaultValue changed from '' to 'null'
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment.points_possible
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.submission.score
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.course_view_option.show_assignment_planning
          defaultValue changed from 'true' to 'null'
          nullable changed from 'true' to 'false'
     my_learning_analytics_dev.course_view_option.show_files_accessed
          defaultValue changed from 'true' to 'null'
          nullable changed from 'true' to 'false'
     my_learning_analytics_dev.course_view_option.show_grade_distribution
          defaultValue changed from 'true' to 'null'
          nullable changed from 'true' to 'false'
     my_learning_analytics_dev.user.sis_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.user.sis_name
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.course.term_id
          defaultValue changed from '1' to 'null'
          nullable changed from 'false' to 'true'
          order changed from '3' to '4'
          type changed from 'BIGINT(20) UNSIGNED' to 'BIGINT(19)'
     my_learning_analytics_dev.submission.user_id
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.user_default_selection.user_id
          nullable changed from 'false' to 'true'
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
     my_learning_analytics_dev.assignment_groups.weight
          type changed from 'CHAR(255 BYTE)' to 'VARCHAR(255 BYTE)'
Missing Foreign Key(s): NONE
Unexpected Foreign Key(s):
     course_view_option_course_id_2efc2b59_fk_course_id(course_view_option[course_id] -> course[id])
     file_access_file_id_id_bee7dfac_fk_file_id(file_access[file_id_id] -> file[id])
     file_access_user_id_id_1b1dbc4e_fk_user_id(file_access[user_id_id] -> user[id])
Changed Foreign Key(s): NONE
Missing Primary Key(s): NONE
Unexpected Primary Key(s):
     PRIMARY on student_dashboard.file_access(id)
Changed Primary Key(s):
     PRIMARY on my_learning_analytics_dev.course(canvas_id)
          columns changed from '[my_learning_analytics_dev.course.canvas_id]' to '[student_dashboard.course.id]'
     PRIMARY on my_learning_analytics_dev.user(id, course_id)
          columns changed from '[my_learning_analytics_dev.user.id, my_learning_analytics_dev.user.course_id]' to '[student_dashboard.user.id]'
     PRIMARY on my_learning_analytics_dev.user_default_selection(id, course_id, user_id, default_view_type)
          columns changed from '[my_learning_analytics_dev.user_default_selection.id, my_learning_analytics_dev.user_default_selection.course_id, my_learning_analytics_dev.user_default_selection.user_id, my_learning_analytics_dev.user_default_selection.default_view_type]' to '[student_dashboard.user_default_selection.id]'
Missing Table(s): NONE
Unexpected Table(s): NONE
Changed Table(s): NONE
Missing View(s): NONE
Unexpected View(s): NONE
Changed View(s): NONE
Liquibase command 'diff' was executed successfully.
```

**Comment by jonespm** [comment]
1) It *sounds* like for these CHAR fields it's fine to go to VARCHAR, we'll just need to convert them. Otherwise we have to write a "hacky" model which I don't think is needed

https://code.djangoproject.com/ticket/9349

2) TIMESTAMP is four bytes vs eight bytes for DATETIME, so sounds like we're fine with additional precision.

3) [We *may* want to look a custom model ](https://stackoverflow.com/a/32516349/3708872)for LONGTEXT, filenames don't need to be LONGTEXT.

4) DOUBLE seems fine over FLOAT

Any thoughts @andrew-gardener ?

**Comment by jonespm** [comment]
Easiest way for some of these fixes looks like using https://github.com/adamchainz/django-mysql

**Comment by jonespm** [comment]
I'm just going to do a conversion for this for now and accept everything in the model. We can change LONGTEXT to TEXT later if we'd like.

**Comment by andrew-gardener** [comment]
@jonespm sorry for taking a while to get back to you on this! 

I've had a look over the schema while doing #540 and I think all your suggestions (1-4) are good. I'll take a look at django-mysql but I don't think we're in a place where we need to worry to much about the exact types (though its a bit weird that Django doesn't give you options to control this directly)

**Comment by jonespm** [comment]
This was fixed in #540
